packages/shared/src/hooks/feed/useFeedUpvotePost.ts

```diff
@@ -27,24 +49,72 @@ export default function useFeedUpvotePost(
   column: number,
   upvoted: boolean,
 ) => Promise<void> {
+  const queryClient = useQueryClient();
   const { user, showLogin } = useContext(AuthContext);
   const { trackEvent } = useContext(AnalyticsContext);
 
-  const { upvotePost, cancelPostUpvote } = useUpvotePost<{
-    id: string;
-    index: number;
-  }>({
-    onUpvotePostMutate: optimisticPostUpdateInFeed(
-      items,
-      updatePost,
-      (post) => ({ upvoted: true, numUpvotes: post.numUpvotes + 1 }),
-    ),
-    onCancelPostUpvoteMutate: optimisticPostUpdateInFeed(
-      items,
-      updatePost,
-      (post) => ({ upvoted: false, numUpvotes: post.numUpvotes - 1 }),
-    ),
-  });
+  const { upvotePost, cancelPostUpvote } =
+    useUpvotePost<UseFeedUpvotePostVariables>({
+      onUpvotePostMutate: optimisticPostUpdateInFeed(
+        items,
+        updatePost,
+        mutationHandlers[upvotePostKey],
+      ),
+      onCancelPostUpvoteMutate: optimisticPostUpdateInFeed(
+        items,
+        updatePost,
+        mutationHandlers[cancelUpvotePostKey],
+      ),
+    });
+
+  useEffect(() => {
```

Comment:
How come the above mutation doesn't catch this?
Seems like a lot of overhead to handle a optimistic update, but might miss the context in this PR

Reply:
So basically the core issue is that upvote data is stored in two (or more) different queries. 

1. For single posts inside `['post', id]`
2. For different feeds (every feed has it's own query keys)

Issue is that `useUpvotePost` mutation's `onSuccess` handlers only modify/optimistically update data inside `1.` query key. 

As for `2.` updating data inside all feed queries would not really be possible since each feed has completely different query key.

Ultimately we would have to refactor our query keys and data on frontend to have a separate query to hold all the upvote data or additional query for each post upvote state. I actually tried going with first idea initially but realised this does not match our API data structure so I found it inefficient to remap all of this for this one user journey.

What I did instead was to inverse the logic and instead of manually trying to update query keys of all related queries I simply subscribe to mutations from active feed itself. Cool bonus is that since squad rely on the feed this automatically works there too.

This way:
- only active/rendered feed actually listens for data and optimistically updates
- no overhead in `useUpvotePost` upvote hook
- we can keep our current data structure and solve this pretty common user journey with minimal changes

FYI @sshanzel I've put this on preview https://preview.app.daily.dev

Reply:
> Moving forward, would you be able to create a reusable hook for such cases? I think we have the same case for bookmarks and comments too.

I would do this on the next task that touches this use case, not sure what would be needed to make the hook reusable at this time. The real fix should be revamping our query keys though.

> This got me thinking, would it be not easier to just pass a callback on the using modal (since it came from the feed) for when we upvote/comment/bookmark the post?

I personally love that this is contained inside a hook that is already specialised for feed upvotes without any additional logic sprinkled through callbacks on other components. This highlights power of global mutations which we get for free in react-query.